File: server/utils/mpesa_parser.py
"""Parser for pasted M-PESA confirmation messages.

Safaricom sends one SMS per transaction, and workers paste one or many of them
into Kipato at once. Only *inbound* money is treated as income: a message about
money the worker sent, paid or withdrew must never become an income entry.

The parser is deliberately tolerant of the wording variants Safaricom uses
("You have received Ksh500.00 from ..." and "Ksh500.00 received from ...") and
still accepts the simple ``YYYY-MM-DD ... 1000.00`` shorthand so hand-typed
records keep working.

`server/utils/mpesa_parser.py` and `frontend/src/lib/mpesa-parser.ts` implement
the same rules; change them together.
"""

import re
from datetime import date, datetime
from typing import List, Optional

# A Safaricom transaction code: 10 alphanumerics, e.g. "RJ12ABC123".
TXN_CODE = r"[A-Z0-9]{10}"

_CODE_RE = re.compile(r"\b(" + TXN_CODE + r")\b(?=\s+Confirmed)", re.IGNORECASE)
_SPLIT_RE = re.compile(r"(?=\b" + TXN_CODE + r"\b\s+Confirmed)", re.IGNORECASE)

_AMOUNT = r"(?:Ksh|KES|Kshs)\s*\.?\s*(\d[\d,]*(?:\.\d{1,2})?)"
# "received Ksh500.00" and "Ksh500.00 received" are both in the wild.
_RECEIVED_THEN_AMOUNT_RE = re.compile(r"receiv(?:ed|e)\s+" + _AMOUNT, re.IGNORECASE)
_AMOUNT_THEN_RECEIVED_RE = re.compile(_AMOUNT + r"\s+receiv(?:ed|e)\b", re.IGNORECASE)

# Money leaving the worker's wallet. Never income.
_OUTBOUND_RE = re.compile(
    r"\b(sent to|paid to|pay bill to|withdraw|bought|buy goods to|"
    r"transferred to|airtime)\b",
    re.IGNORECASE,
)
_INBOUND_RE = re.compile(r"\breceiv(?:ed|e)\b", re.IGNORECASE)

_DMY_RE = re.compile(r"\bon\s+(\d{1,2})/(\d{1,2})/(\d{2,4})\b", re.IGNORECASE)
_ISO_RE = re.compile(r"\b(\d{4})-(\d{2})-(\d{2})\b")

_SENDER_RE = re.compile(
    r"\bfrom\s+([A-Za-z][A-Za-z.'\- ]{1,60}?)"
    r"(?=\s+(?:\+?254\d{9}|0\d{9}|\d{10,12})\b|\s+on\b|\s*[.,]|$)",
    re.IGNORECASE,
)

# Fallback for hand-typed lines: "2026-09-01 Received 1000.00 from ...".
_PLAIN_RE = re.compile(r"(?P<date>\d{4}-\d{2}-\d{2}).*?(?P<amount>\d[\d,]*\.\d{2})")

MAX_AMOUNT = 1000000


def _to_amount(raw: str) -> Optional[float]:
    try:
        value = float(raw.replace(",", ""))
    except (TypeError, ValueError):
        return None
    if value <= 0 or value > MAX_AMOUNT:
        return None
    return value


def _to_date(day: int, month: int, year: int) -> Optional[date]:
    if year < 100:
        year += 2000
    try:
        return date(year, month, day)
    except ValueError:
        return None


def _find_date(segment: str) -> Optional[date]:
    match = _DMY_RE.search(segment)
    if match:
        # Safaricom formats Kenyan dates as day/month/year.
        parsed = _to_date(int(match.group(1)), int(match.group(2)), int(match.group(3)))
        if parsed:
            return parsed
    match = _ISO_RE.search(segment)
    if match:
        return _to_date(int(match.group(3)), int(match.group(2)), int(match.group(1)))
    return None


def _find_amount(segment: str) -> Optional[float]:
    # Anchor on the "received" keyword so the closing balance is never mistaken
    # for the transaction amount.
    for pattern in (_RECEIVED_THEN_AMOUNT_RE, _AMOUNT_THEN_RECEIVED_RE):
        match = pattern.search(segment)
        if match:
            amount = _to_amount(match.group(1))
            if amount is not None:
                return amount
    return None


def _find_sender(segment: str) -> Optional[str]:
    match = _SENDER_RE.search(segment)
    if not match:
        return None
    sender = " ".join(match.group(1).split()).strip(" .,-")
    return sender[:120] or None


def _split_messages(raw_text: str) -> List[str]:
    segments: List[str] = []
    for block in re.split(r"[\r\n]+", raw_text):
        block = block.strip()
        if not block:
            continue
        # Several SMS pasted onto one line still split on their transaction code.
        for part in _SPLIT_RE.split(block):
            part = " ".join(part.split())
            if part:
                segments.append(part)
    return segments


def parse_mpesa_text(raw_text: str) -> List[dict]:
    """Extract inbound transactions from pasted M-PESA text.

    Returns a list of ``{"code", "amount", "date", "sender", "raw"}`` dicts,
    ordered as they appear in the text and de-duplicated by transaction code.
    """
    if not raw_text:
        return []

    entries: List[dict] = []
    seen_codes = set()

    for segment in _split_messages(raw_text):
        if _OUTBOUND_RE.search(segment) and not _INBOUND_RE.search(segment):
            continue

        amount = _find_amount(segment)
        if amount is None and _OUTBOUND_RE.search(segment):
            # An outbound message that merely mentions "received" elsewhere.
            continue

        entry_date = _find_date(segment)

        if amount is None or entry_date is None:
            match = _PLAIN_RE.search(segment)
            if not match:
                continue
            amount = _to_amount(match.group("amount"))
            entry_date = _find_date(match.group("date"))
            if amount is None or entry_date is None:
                continue

        code_match = _CODE_RE.search(segment)
        code = code_match.group(1).upper() if code_match else None
        if code and code in seen_codes:
            continue
        if code:
            seen_codes.add(code)

        entries.append({
            "code": code,
            "amount": amount,
            "date": entry_date,
            "sender": _find_sender(segment),
            "raw": segment[:500],
        })

    return entries


def parse_iso_date(value: str) -> Optional[date]:
    try:
        return datetime.strptime(value, "%Y-%m-%d").date()
    except (TypeError, ValueError):
        return None

File: frontend/src/lib/mpesa-parser.ts
import type { ParsedMpesaEntry } from '../types'

/**
 * Parser for pasted M-PESA confirmation messages.
 *
 * Only *inbound* money counts as income: a message about money the worker sent,
 * paid or withdrew must never become an income entry.
 *
 * This mirrors `server/utils/mpesa_parser.py`; change them together.
 */

// A Safaricom transaction code: 10 alphanumerics, e.g. "RJ12ABC123".
const TXN_CODE = '[A-Z0-9]{10}'

const CODE_RE = new RegExp(`\\b(${TXN_CODE})\\b(?=\\s+Confirmed)`, 'i')
const SPLIT_RE = new RegExp(`(?=\\b${TXN_CODE}\\b\\s+Confirmed)`, 'i')

const AMOUNT = '(?:Ksh|KES|Kshs)\\s*\\.?\\s*(\\d[\\d,]*(?:\\.\\d{1,2})?)'
// "received Ksh500.00" and "Ksh500.00 received" are both in the wild.
const RECEIVED_THEN_AMOUNT_RE = new RegExp(`receiv(?:ed|e)\\s+${AMOUNT}`, 'i')
const AMOUNT_THEN_RECEIVED_RE = new RegExp(`${AMOUNT}\\s+receiv(?:ed|e)\\b`, 'i')

// Money leaving the worker's wallet. Never income.
const OUTBOUND_RE =
  /\b(sent to|paid to|pay bill to|withdraw|bought|buy goods to|transferred to|airtime)\b/i
const INBOUND_RE = /\breceiv(?:ed|e)\b/i

const DMY_RE = /\bon\s+(\d{1,2})\/(\d{1,2})\/(\d{2,4})\b/i
const ISO_RE = /\b(\d{4})-(\d{2})-(\d{2})\b/

const SENDER_RE =
  /\bfrom\s+([A-Za-z][A-Za-z.'\- ]{1,60}?)(?=\s+(?:\+?254\d{9}|0\d{9}|\d{10,12})\b|\s+on\b|\s*[.,]|$)/i

// Fallback for hand-typed lines: "2026-09-01 Received 1000.00 from ...".
const PLAIN_RE = /(\d{4}-\d{2}-\d{2}).*?(\d[\d,]*\.\d{2})/

const MAX_AMOUNT = 1_000_000

function toAmount(raw: string): number | null {
  const value = Number(raw.replace(/,/g, ''))
  if (!Number.isFinite(value) || value <= 0 || value > MAX_AMOUNT) return null
  return value
}

function toIsoDate(day: number, month: number, year: number): string | null {
  const fullYear = year < 100 ? year + 2000 : year
  if (month < 1 || month > 12 || day < 1 || day > 31) return null

  const parsed = new Date(Date.UTC(fullYear, month - 1, day))
  // Rejects the likes of 31 February, which Date would otherwise roll forward.
  if (
    parsed.getUTCFullYear() !== fullYear ||
    parsed.getUTCMonth() !== month - 1 ||
    parsed.getUTCDate() !== day
  ) {
    return null
  }
  return parsed.toISOString().slice(0, 10)
}

function findDate(segment: string): string | null {
  const dmy = DMY_RE.exec(segment)
  if (dmy) {
    // Safaricom formats Kenyan dates as day/month/year.
    const parsed = toIsoDate(Number(dmy[1]), Number(dmy[2]), Number(dmy[3]))
    if (parsed) return parsed
  }
  const iso = ISO_RE.exec(segment)
  if (iso) return toIsoDate(Number(iso[3]), Number(iso[2]), Number(iso[1]))
  return null
}

function findAmount(segment: string): number | null {
  // Anchor on the "received" keyword so the closing balance is never mistaken
  // for the transaction amount.
  for (const pattern of [RECEIVED_THEN_AMOUNT_RE, AMOUNT_THEN_RECEIVED_RE]) {
    const match = pattern.exec(segment)
    if (match) {
      const amount = toAmount(match[1])
      if (amount !== null) return amount
    }
  }
  return null
}

function findSender(segment: string): string | null {
  const match = SENDER_RE.exec(segment)
  if (!match) return null
  const sender = match[1].split(/\s+/).join(' ').replace(/^[.,\-\s]+|[.,\-\s]+$/g, '')
  return sender ? sender.slice(0, 120) : null
}

function splitMessages(rawText: string): string[] {
  const segments: string[] = []
  for (const block of rawText.split(/[\r\n]+/)) {
    const trimmed = block.trim()
    if (!trimmed) continue
    // Several SMS pasted onto one line still split on their transaction code.
    for (const part of trimmed.split(SPLIT_RE)) {
      const normalised = part.split(/\s+/).join(' ').trim()
      if (normalised) segments.push(normalised)
    }
  }
  return segments
}

export function parseMpesaText(rawText: string): ParsedMpesaEntry[] {
  if (!rawText) return []

  const entries: ParsedMpesaEntry[] = []
  const seenCodes = new Set<string>()

  for (const segment of splitMessages(rawText)) {
    const outbound = OUTBOUND_RE.test(segment)
    if (outbound && !INBOUND_RE.test(segment)) continue

    let amount = findAmount(segment)
    if (amount === null && outbound) continue

    let date = findDate(segment)

    if (amount === null || date === null) {
      const plain = PLAIN_RE.exec(segment)
      if (!plain) continue
      amount = toAmount(plain[2])
      date = findDate(plain[1])
      if (amount === null || date === null) continue
    }

    const codeMatch = CODE_RE.exec(segment)
    const code = codeMatch ? codeMatch[1].toUpperCase() : null
    if (code) {
      if (seenCodes.has(code)) continue
      seenCodes.add(code)
    }

    entries.push({
      code,
      date,
      amount,
      sender: findSender(segment),
      rawMatch: segment.slice(0, 500),
    })
  }

  return entries
}
